Proposal I - compatibilities extension in image index platform and image manifest config

ChaoyiHuang opened the pull request:
This proposal is born from proposal A to H.

The purpose of the proposal, based on the maintainers feedback, is to address the concern maintaining standalone artifact in OCI.

ChaoyiHuang committed:
Proposal I -  - compatibilities extention in platform object

This proposal is born from proposal A to H.

The purpose of the proposal, based on the maintainers feedback,
is to address the concern maintaining standalone artifact in OCI.

Signed-off-by: joehuang <joehuang.sweden@gmail.com>

mfranczy commented on docs/proposals/PROPOSAL_I.md:
```diff
@@ -0,0 +1,235 @@
+# Proposal I - compatibilities extention in platform object
```
s/extention/extension/g

ChaoyiHuang replied:
OK, update in next commit

mfranczy commented on docs/proposals/PROPOSAL_I.md:
````diff
@@ -0,0 +1,235 @@
+}
+```
+
+### Config Extention
````
Instead of extending the config, I think, it's more reasonable to add a new schema for new image compatibility media type under image spec.
https://github.com/opencontainers/image-spec/blob/main/media-types.md#oci-image-media-types

ChaoyiHuang replied:
it could be another option. will add a new proposal to short the path

mfranczy replied:
We should not bring image config to the index level. The descriptor idea makes only sense to me when we create a new compatibility media type. I think this proposal and J should be combined to one.

vsoch replied:
>  I think this proposal and J should be combined to one.

+1 - it would be easier to have one WIP proposal.

ChaoyiHuang replied:
Just explained to Marcin, it's two different proposals. I will update the api call path in the proposal how to get the compatibilities

ChaoyiHuang replied:
some stuff to help understand proposal I and proposal J: https://docs.google.com/presentation/d/1w_MNe2qEJH6Izcq61IABPfx_C9cUY1405j_FgCs8e1Q/edit?usp=sharing

mfranczy replied:
Yup, indeed those are two different proposals, although as I mentioned, the proposal violates the existing media types relations: https://github.com/opencontainers/image-spec/blob/main/media-types.md#relations

mfranczy replied:
Closing this thread as there is a new proposal.

mfranczy commented:
Definetely the idea of adding a new `compatibilities` field in the index that is a descriptor is interesting.
As mentioned in the comment, if we go that way we should create a new media type `application/vnd.oci.image.compatibility.v1+json` and provide its own properties (probably the one we wanted for artifact).

I also like the idea because compatibility would be under the image spec. If container runtimes want to support it they would not have a lot of additional API calls.

vsoch commented on docs/proposals/PROPOSAL_I.md:
```diff
@@ -0,0 +1,235 @@
+
+This proposal is born from proposal A to H.
+
+The purpose of the proposal, based on the maintainers feedback, is to address the concern maintaining standalone artifact in OCI.
```
```suggestion
The purpose of the proposal, based on the maintainers feedback, is to address the concern maintaining a standalone artifact in OCI.
```

ChaoyiHuang replied:
ok, will update in next commit

vsoch commented on docs/proposals/PROPOSAL_I.md:
```diff
@@ -0,0 +1,235 @@
+
+The purpose of the proposal, based on the maintainers feedback, is to address the concern maintaining standalone artifact in OCI.
+
+## image spec platform object extention
```
```suggestion
## image spec platform object extension
```

ChaoyiHuang replied:
yeah, spelling issue. will fix

vsoch commented on docs/proposals/PROPOSAL_I.md:
```diff
@@ -0,0 +1,235 @@
+
+### add platform.compatibilities descriptor
+
+Add a new descriptor in platform object in [image index](https://github.com/opencontainers/image-spec/blob/main/image-index.md#image-index-property-descriptions).
```
Doesn't this still imply another artifact, it's just linked to differently, directly from a manifest list?

ChaoyiHuang replied:
The idea here is to address our concerns in current OCI specifications, avoid introduing a new standalone spec ( maintainers worried about new standard to maintain). so just add new descriptor in  manifest entry's platform objector in image index, the descriptor will point to a manifest, in which the compatibilities could reside in config(the current commit), or in a new mediatype layer(another option).
 
it can be handled quite similar to artifact, but no need to use referrers api anymore. 

the compatibilities now can be decoupled from container lifecycle，and optional for runtime engine, runtime engine can be configured to use the compatibilities or not, in sync or async order, it's easy for runtime engine to evolve.

vsoch commented on docs/proposals/PROPOSAL_I.md:
```diff
@@ -0,0 +1,235 @@
+
+    - **`features`** _array of strings_
+
+    - **`compatibilities`** _[descriptor](https://github.com/opencontainers/image-spec/blob/main/descriptor.md)_
```
This is confusing because now I'm thinking the field "compatibilities" is part of a descriptor.

ChaoyiHuang replied:
maybe another term would be better. your suggestion?

vsoch commented on docs/proposals/PROPOSAL_I.md:
```diff
@@ -0,0 +1,235 @@
+
+## Compatibilities Manifest
+
+The compatibilities manifest is a special image manifest, contains only config and no blob layer.
```
So this is another link / redirection to get to the compatibilities? If I understand:

image manifest list -> descriptor ->  image manifest -> config (which has them)?

ChaoyiHuang replied:
> So this is another link / redirection to get to the compatibilities? If I understand:
> 
> image manifest list -> descriptor -> image manifest -> config (which has them)?

yes, this is the path to get the compatibilities content.

will submit another proposal to short the path by adding a new mediatype, as commented by Marcin

vsoch commented on docs/proposals/PROPOSAL_I.md:
```diff
@@ -0,0 +1,235 @@
+
+### Config Extention
+
+The config object will be extened with compatibilities field.
```
Does this mean we need to do diffs and similar if you are combining layers from multiple images / multiple configs?

ChaoyiHuang replied:
no need. in this config, only compatibilities is valid content.

vsoch commented on docs/proposals/PROPOSAL_I.md:
````diff
@@ -0,0 +1,235 @@
+#### Example - image is compatible with different hardware combinations
+
+```json
+{
````
This looks like just slightly changing the format / structure to look like a config over an artifact. There are still many jumps to get here - I'm not sure why this approach is better? It has more down sides, like requiring the person own the image / registry URI to be able to update or define the compatibilities. And then the issue of changing digests.

ChaoyiHuang replied:
for jump, it's similar as artifact.

> It has more down sides, like requiring the person own the image / registry URI to be able to update or define the compatibilities. And then the issue of changing digests.

it's quite easy to update compatibilites, just replace the descriptor in the image index manifest entry. and the image manifest and blob layes itself doesn't need to update (no rebuild and redistribution), the image manifest digest didnot change. it's very flexible while keep current flow unchanged.

ChaoyiHuang committed:
compatibilities extension in image index platform and image manifest config

This proposal is born from proposal A to H.

Based on the maintainers feedback, The purpose of the proposal,
is to address the concern maintaining standalone artifact in OCI.
Spec update is done on image index platform and image manifest config.

Signed-off-by: joehuang <joehuang.sweden@gmail.com>

sudo-bmitch commented on docs/proposals/PROPOSAL_I.md:
```diff
@@ -0,0 +1,235 @@
+
+### add platform.compatibilities descriptor
+
+Add a new descriptor in platform object in [image index](https://github.com/opencontainers/image-spec/blob/main/image-index.md#image-index-property-descriptions).
```
Same issue that I raised on `J`: I have concerns that this descriptor would not be known to existing registries and client tooling. As a result, registries would accept an incomplete manifest, or silently delete the compatibility blob in a later GC cycle. And client tooling to copy images would fail to include the compatibility blobs.

mfranczy replied:
@sudo-bmitch is right, so far I confirmed that with regclient and zot registry.
For this proposal we would have to add a descriptor support to regclient and zot.
Didn't test other registries and tools yet.

sudo-bmitch reviewed (changes requested):
Putting this on hold until we can discuss the risk of unexpected data loss.
